vault/identity_store_oidc_provider.go

```diff
@@ -277,7 +278,163 @@ func oidcProviderPaths(i *IdentityStore) []*framework.Path {
 			HelpSynopsis:    "Query OIDC configurations",
 			HelpDescription: "Query this path to retrieve the configured OIDC Issuer and Keys endpoints, response types, subject types, and signing algorithms used by the OIDC backend.",
 		},
+		{
+			Pattern: "oidc/provider/" + framework.GenericNameRegex("name") + "/.well-known/keys",
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeString,
+					Description: "Name of the provider",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.ReadOperation: i.pathOIDCReadProviderPublicKeys,
+			},
+			HelpSynopsis:    "Retrieve public keys",
+			HelpDescription: "Returns the public portion of keys for a named OIDC provider. Clients can use them to validate the authenticity of an ID token.",
+		},
+	}
+}
+
+// TODO: load clients into memory (go-memdb) to look this up
+func (i *IdentityStore) allClients(ctx context.Context, s logical.Storage) (*[]client, error) {
+	clientNames, err := s.List(ctx, clientPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var clients []client
+	for _, name := range clientNames {
+		entry, err := s.Get(ctx, clientPath+name)
+		if err != nil {
+			return nil, err
+		}
+		if entry == nil {
+			continue
+		}
+
+		var client client
+		if err := entry.DecodeJSON(&client); err != nil {
+			return nil, err
+		}
+		clients = append(clients, client)
+	}
+
+	return &clients, nil
+}
+
+// TODO: load clients into memory (go-memdb) to look this up
+func (i *IdentityStore) clientByID(ctx context.Context, s logical.Storage, id string) (*client, error) {
+	clients, err := i.allClients(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, client := range *clients {
+		if client.ClientID == id {
+			return &client, nil
+		}
 	}
+
+	return nil, nil
+}
+
+// keyIDsReferencedByTargetClientIDs returns a slice of key IDs that are
+// referenced by the clients' targetIDs.
+// If targetIDs contains "*" then the IDs for all public keys are returned.
+func (i *IdentityStore) keyIDsReferencedByTargetClientIDs(ctx context.Context, s logical.Storage, targetIDs []string) ([]string, error) {
+	var keyNames []string
```

Comment:
I wonder if we can simplify this a bit. I think we could get duplicate keys published if the `targetIDs = ["key-id-1", "*"]` given that keyNames isn't a set (map in Go).

Leaving an approach below, but feel free to implement this how you'd prefer:
```go
// This could be a map[string]bool to ensure set behavior (no dup keys), but doesn't have to be.
var keyNames []string

// Get all key names referenced by clients if wildcard "*" in target client IDs
if strutil.StrListContains(targetIDs, "*") {
	clients, err := i.allClients(ctx, s)
	if err != nil {
		return nil, err
	}

	for _, client := range clients {
		keyNames = append(keyNames, client.Key)
	}
}

// Otherwise, get the key names referenced by each target client ID
if len(keyNames) == 0 {
	for _, clientID := range targetIDs {
		client, err := i.clientByID(ctx, s, clientID)
		if err != nil {
			return nil, err
		}

		if client != nil {
			keyNames = append(keyNames, client.Key)
		}
	}
}

// Collect the key IDs
...
```

Reply:
We could return right away after collecting the keys for the glob case (since that contains the key for all other clients specified) and remove the need for `if len(keyNames) == 0 {...}`.

Reply:
There is some more processing of `keyNames` after this suggestion that I shared, so I don't think we can immediately return after collecting the keys in the glob case.

Reply:
Good point and I like the idea of using `map[string]bool`